### PR TITLE
Track U at bottom of system

### LINKED ISSUE #22: Track U at the bottom of the system (500 cm)
Request from @EmiliaJarochowska 

We want to know how much solid is moved out of the system. U is needed for this.

### PR DESCRIPTION
**Pull request by HannoSpreeuw**
Fixes #22 

Also added live plotting (although unrelated to that issue). Final graph is stored as `Final_distributions.pdf` in `stored_folder`.

The values of U at the bottom of the system and the corresponding times as a fraction of the total integration time span - which, in turn, is expressed in units of T* - are recorded in a single Pandas dataframe though py-pde's DataTracker class. This class is instantiated calling a function which has access to the state of the system during integration. This function rederives U at intervals set in the parameters.Tracker dataclass.

Now, since the Pandas dataframe is stored as an hdf5 file, the simplest way to read in this hdf5 file into a Numpy array and plot it is something like this:

```
import numpy as np
import pandas as pd
pandas_df = pd.read_hdf("File_just_created.h5")
U_at_bottom_and_times = pandas_df.to_numpy()
U_at_bottom_and_times.shape
(101, 2)
import pylab as pyl
pyl.plot(U_at_bottom_and_times[:, 0], U_at_bottom_and_times[:, 1], "ro")
pyl.show()
```

### CODE REVIEW

**Comment by EmiliaJarochowska** [comment]
`pipenv install` gives errors (attached) and stops. This doesn't happen on `main`.
[pipenv_error.txt](https://github.com/astro-turing/Integrating-diagenetic-equations-using-Python/files/12471966/pipenv_error.txt)

**Comment by EmiliaJarochowska** [comment]
This was originally requested by @NiklasHohmann so I added him to take a look.

**Comment by NiklasHohmann** [comment]
Same as for @EmiliaJarochowska , `pipenv install` throws errors

**Comment by HannoSpreeuw** [comment]
Does deleting Pipfile.lock solve the problem?

**Comment by HannoSpreeuw** [comment]
And does this also happen from a fresh git clone?

**Comment by EmiliaJarochowska** [comment]
It _had been_ solved after a fresh git clone but just running a new `pipenv install` in this folder again brings back the errors.
[https://mastodon.social/@jk/110769222900259244](https://mastodon.social/@jk/110769222900259244)

**Comment by HannoSpreeuw** [comment]
Does the error also reoccur when you first delete Pipfile.lock?

I could consider stopping to git track that file.

**Comment by HannoSpreeuw** [comment]
From a discussion with @NiklasHohmann it seems that me adding `pandas` and `tables` to the Pipfile has caused problems. But they are needed for writing results to hdf5 files. These results are collected in a Pandas dataframe.

From a first glance at your [pipenv_error.txt](https://github.com/astro-turing/Integrating-diagenetic-equations-using-Python/files/12471966/pipenv_error.txt), it may be that your problems have the same cause, but I am not 100% sure.

**Comment by HannoSpreeuw** [comment]
Btw, how does a `poetry install` work for you?

**Comment by EmiliaJarochowska** [comment]
I thought `pipenv install --skip-lock` would also solve the problem but it doesn't. 
[pipenv_errors.txt](https://github.com/astro-turing/Integrating-diagenetic-equations-using-Python/files/12484712/pipenv_errors.txt)
[poetry_errors.txt](https://github.com/astro-turing/Integrating-diagenetic-equations-using-Python/files/12484714/poetry_errors.txt)

**Comment by HannoSpreeuw** [comment]
Perhaps I have to update the instructions for a pipenv install now that I have added these two packages. That would require help from a conda Python 3.11 environment:
```
What you can do, is a pipenv install while using packages already installed in 
some conda environment. Let's say a Python 3.11 environment which conda refers
to as py311, but you have obviously chosen whatever name for that conda 
environment that you like.

Now in your conda py311 environment install pandas and tables:

conda install pandas
conda install -c anaconda pytables

Going back to your conda base environment, you can do a pipenv install
while making use of the conda py311 installed packages. 

pipenv --python=/path/to/python --site-packages

You need to insert the path to your python executable in your conda 
py311 environment.
```
Instructions from [here](https://docs.pipenv.org/advanced/)

**Comment by EmiliaJarochowska** [comment]
That requires using conda. Is this not too restrictive a requirement?

**Comment by HannoSpreeuw** [comment]
@EmiliaJarochowska Thanks for trying `poetry` also. That message is clear:

```
Note: This error originates from the build backend, and is likely not a
problem with poetry but with tables (3.8.0) not supporting PEP 517 builds.
You can verify this by running 'pip wheel --use-pep517 "tables
(==3.8.0)"'.
```

But it is yet unclear to me how to fix it.

**Comment by HannoSpreeuw** [comment]
@EmiliaJarochowska Wrt the additional requirement of Conda: that is an issue indeed. Not sure how to proceed.

**Comment by HannoSpreeuw** [comment]
Okay, [SO shows an alternative solution, i.e. alternative to conda, but related to a different error message](https://stackoverflow.com/questions/76780968/pytables-install-with-python-3-11-fails-on-macos-m1), by installing PyTables from source:
```
pip install git+https://github.com/PyTables/PyTables
````
I could add that to `pyproject.toml` and to `Pipfile`, to see if that works for you. But your issue may be unrelated.

**Comment by EmiliaJarochowska** [comment]
I tried locally as well, although not sure if correctly, and it didn't work either. Trying it on linux, but I have some issues with pip so I didn't get to that step yet.

**Comment by HannoSpreeuw** [comment]
Thanks for trying.

**Comment by HannoSpreeuw** [comment]
CI gave no problems, so I guess the `pipenv install` gave no problems on the OS used by GH Actions (probably Linux).

**Comment by EmiliaJarochowska** [comment]
The problems with running are, within certain constraints, resolved. So I wanted to check what values we get: Whether anything is moving across the boundary and in what direction.
You can see an example from a run here:
https://github.com/MindTheGap-ERC/LMA_utils/blob/main/Track_U_at_bottom.ipynb

The U values saved in this branch are different from those calculated using Eq. 46 so I wonder if I made a mistake. @NiklasHohmann could you please double check? Most likely my calculation is wrong but I would like to clarify this before closing the PR.

Having all keys in a hdf5 file is amazing for keeping track of run outputs. 
Next step is having them somewhere we can access them. I have put this run at https://osf.io/ews27/ where it can be downloaded directly into a notebook. 
Is this useful? If you think you need to be saving runs where we can access them, I can give you write access.

**Comment by EmiliaJarochowska** [comment]
Thank you very much for checking. 
I tried a run with higher porosities and used that to compare U that is saved into the file, what the value would be calculated using the steps in `marlpde` and what it would be according to Eq. 46. In [the same place](https://github.com/MindTheGap-ERC/LMA_utils/blob/main/Track_U_at_bottom.ipynb).

**U calculated from Eq. 46 = 110.74, U exported into hdf5 = 12.54, U calculated in marlpde = 11.97**

_U_ is dimensionless _u_, meaning 𝑈 = 𝑢/𝑆. S is 0.1, so _u_ from Eq. 46 would be 11.074 which is indeed quite close to that coming out from `marlpde`. I can't pinpoint if Eq. 46 should be _u_ not _U_, i.e. the scaling factor is missing in either direction.

The value of _u_ is what we'd be happy with, i.e. positive, but hard to tell if we can trust it since the results of the run are clearly wrong.

Our discussions on Teams illustrated the need for comparing outputs from various runs systematically. It would be handy to render this notebook into a static document with values of different runs. I might switch from Jupyter to Quarto for this. Or add entangled tags into `marlpde`; that would be cool but I struggled a bit with rendering with it. I clearly should have attended @jhidding's talk as I don't know what the best way is here.

**Comment by HannoSpreeuw** [comment]
Clearly wrong....? So that refers to your Teams message, showing a graph depicting U gradually dropping to a small positive number (0.094) after 16T*? 
Why is that not possible?

**Comment by HannoSpreeuw** [comment]
> U calculated from Eq. 46 = 110.74, U exported into hdf5 = 12.54, U calculated in marlpde = 11.97

So trying to explain the factor 10 difference....
I don't yet understand why you would do an extra division by S.
You have
```
Vscale = S
P_32 = beta/Vscale
betasV = P_32
k = betasV * phi ** 3 / (1 - phi) ** 2
k = k * (1 - np.exp(-10 * (1 - phi) / phi))
```
So effectively
```
betasV = beta/S
```
Why the extra divison by S?

**Comment by EmiliaJarochowska** [comment]
Oh no. Of course you're right. There's no division by S in k. I rewrote the equation and still didn't see it 😔
I rewrote [the calculation of k directly from Fortran](https://github.com/astro-turing/lheureux/blob/6fd4acc010aad2dbc82d23aac16546ba751beed4/src/lheureux.f#L312-L313).

After your correction:
U calculated from Eq. 46 = 11.97, U exported into hdf5 = 12.54, U calculated in marlpde = 11.97

Also for the other run I made, with Phi0 = PhiIni = 0.75 it's much closer. And if we add units:
_u_ calculated from Eq. 46 = 0.002 cm/a, _u_ exported into hdf5 = 0.009 cm/a, _u_ calculated in marlpde = 0.002 cm/a

Is that a lot? For carbonates most accumulation rates lie between 0.0001 and 1 cm/a ([Kemp & Sadler 2014](https://onlinelibrary.wiley.com/doi/abs/10.1111/sed.12112)). These are called sedimentation rates in the paper, but that's wrong: sedimentation is what comes on the surface with its initial porosity, equaling to production in CarboKitten, but what is measured there and what we try to obtain is deposition rate, after compaction. So these data should be comparable, even if they're not labeled correctly in the paper. Maybe @NiklasHohmann will have more insights on this?

TL;DR the values are realistic. But what's saved in the hdf5 is slightly different than what is calculated in marlpde, although that calculation is correct. And probably there was another error in the Fortran code 😬
I ran out of battery to hotspot myself so that's it for now.

**Comment by NiklasHohmann** [comment]
RE sedimentation rates at the surface vs. compacted export at the bottom of the system: I calculated the proportion of material lost in Fortran, and got to values of around 80 %, so rates of export at the bottom can be one order of magnitude lower than input at the surface. Physically the explanation is that the material is dissolved and exported via the pore fluid.
RE deposition vs. sedimentation rate: There seems to be no consensus in the usage of these terms. Because of the underlying scaling laws when time scales of observations are varied (e.g., the Sadler effect) and the lack of erosion in our model I'd be cautious when comparing our model values (which are instantaneous rates) with rate compilations from deep time. Maybe it is worth to extract the mass balance of CA, AR & TE from the model outputs. This will give us a clue whether
 * the degree of dissolution in the system is plausible, and 
 * whether the difference between rates at the surface and bottom is driven by compaction or dissolution

**Comment by EmiliaJarochowska** [comment]
> Clearly wrong....? So that refers to your Teams message, showing a graph depicting U gradually dropping to a small positive number (0.094) after 16T*?

Sorry, that was indeed unclear. "Clearly wrong" referred to the porosity at the bottom. The system relies on high compressibility, but it doesn't seem to drop very much. I had in mind that in the model and in Fortran $\phi_{\inf} = \epsilon = 1e^{-2}$ and that intuitively makes sense, but in marlpde porosity remains high through the grid and through long runs, e.g.:

image
so u and phi seem to have very similar development. What happens at 16 T*?

image

But L'Heureux also reports only a small drop in porosity at the bottom in his Fig. 5, which is inconsistent with his own PhiInf. However, that's for very short runs. One would expect continuous compaction until an equilibrium is reached.

**Comment by HannoSpreeuw** [comment]
> I had in mind that in the model and in Fortran $\phi_{inf} = 0.01$

Ah, do you mean that you interpreted $\phi_{inf}$ as the porosity reached after a long time?
As you noted, in none of the L'Heureux figures such a low porosity is ever reached.

Not sure how to interpret $\phi_{inf}$ .... first used in equation (23), while referring to [Boudreau and Bennett (1999)](https://www.researchgate.net/publication/235743113_New_rheological_and_porosity_equations_for_steady-state_compaction?_sg=HnXMrQ4k0mOJGsZV8KcJ4fXJssqmPziTu9kPO39O5R0cUnSQF3NBop7u9P24xdVC2RJJUz3slPe4cdM&_tp=eyJjb250ZXh0Ijp7ImZpcnN0UGFnZSI6InB1YmxpY2F0aW9uIiwicGFnZSI6Il9kaXJlY3QifX0) (among others).
Below their equation (5) it says:
```
$\phi_{inf}$ is the apparently constant porosity reached at some depth....
```
Is it possible that they mean a depth much deeper than the maximum depth of 5m that we are considering?

**Comment by HannoSpreeuw** [comment]
> But what's saved in the hdf5 is slightly different than what is calculated in marlpde, although that calculation is correct.

I am puzzled by this. What should have been saved is calculated in [this code snippet](https://github.com/MindTheGap-ERC/reactive-transport-model-for-limestone-marl-sequences/blob/c2aa90b11c159e678b7424c2c133b10510d000c8/marlpde/LHeureux_model.py#L105):
```
        Phi = state.data[4][-1]
        F = 1 - np.exp(10 - 10 / Phi)
        one_minus_Phi = 1 - Phi
        U = self.presum + self.rhorat * Phi ** 3 * F /one_minus_Phi
        return {"U at bottom": U}
```
So if we compare this calculation with the ["regular" calculation of U in marlpde](https://github.com/MindTheGap-ERC/reactive-transport-model-for-limestone-marl-sequences/blob/c2aa90b11c159e678b7424c2c133b10510d000c8/marlpde/LHeureux_model.py#L349) we should know the exact value of the porosity that was used.

**Comment by HannoSpreeuw** [comment]
With this last commit, the Pytables dependency has been removed, for both the pipenv and poetry installs, since this casused problems for both Mac OS and Windows. 
We could get around the Pytables dependency by converting the pandas dataframe, containing the data for U at the bottom, to a Numpy array and store that in an HDF5 file.
We decided to use the HDF5 file already created for the five fields, i.e. LMAHeureuxPorosityDiff.hdf5.
We added a separate group called "U", containing a dataset called "U_at_bottom".

The way to retrieve the data is as follows:

```
import h5py
hf = h5py.File(some_path_with_trailing_slash + "LMAHeureuxPorosityDiff.hdf5", "r")

# Show all the entries for metadata
hf.attrs.keys()

# Show all the groups
hf.keys()
----> <KeysViewHDF5 ['U', 'data', 'times']>

# Retrieve the U at the bottom data
U = hf.get('U')
U.keys()
----> <KeysViewHDF5 ['U_at_bottom']>
U_at_bottom = U.get('U_at_bottom')

# Plot U at the bottom as a function of time.
import pylab as pyl
pyl.plot(U_at_bottom[:, 0], U_at_bottom[:, 1], "ro")
pyl.show()

# Turn U_at_bottom into Numpy array
import numpy as np
U_at_bottom = np.array(U_at_bottom)
```

**Comment by HannoSpreeuw** [comment]
> Oh no. Of course you're right. There's no division by S in k. I rewrote the equation and still didn't see it 😔 I rewrote [the calculation of k directly from Fortran](https://github.com/astro-turing/lheureux/blob/6fd4acc010aad2dbc82d23aac16546ba751beed4/src/lheureux.f#L312-L313).
> 
> After your correction: U calculated from Eq. 46 = 11.97, U exported into hdf5 = 12.54, U calculated in marlpde = 11.97
> 
> Also for the other run I made, with Phi0 = PhiIni = 0.75 it's much closer. And if we add units: _u_ calculated from Eq. 46 = 0.002 cm/a, _u_ exported into hdf5 = 0.009 cm/a, _u_ calculated in marlpde = 0.002 cm/a
> 
> Is that a lot? For carbonates most accumulation rates lie between 0.0001 and 1 cm/a ([Kemp & Sadler 2014](https://onlinelibrary.wiley.com/doi/abs/10.1111/sed.12112)). These are called sedimentation rates in the paper, but that's wrong: sedimentation is what comes on the surface with its initial porosity, equaling to production in CarboKitten, but what is measured there and what we try to obtain is deposition rate, after compaction. So these data should be comparable, even if they're not labeled correctly in the paper. Maybe @NiklasHohmann will have more insights on this?
> 
> TL;DR the values are realistic. But what's saved in the hdf5 is slightly different than what is calculated in marlpde, although that calculation is correct. And probably there was another error in the Fortran code 😬 I'm sorry about the mistake. I ran out of battery to hotspot myself so that's it for now.

Okay....trying to reproduce that difference. For which Phi0 & PhiIni did you find values of U at the bottom of 12.54 (saved) and 11.97 (marlpde) respectively? 

You derived the marlpde value of U at the bottom by hand, as shown in your notebook.

If I run marlpde in a debugger with suitable breakpoints, I do find agreement between U at the bottom as derived in `def track_U_at_bottom(self, state, t):` and in `U = self.presum + self.rhorat * Phi ** 3 * F /one_minus_Phi` in `def evolution_rate(self, state, t=0):` (using the slow Numpy backend, which is more suitable for debugging), at least after one time step.

**Comment by EmiliaJarochowska** [comment]
That was for Phi0 = PhiIni = 0.8. You can see in the plot at the bottom.

**Comment by HannoSpreeuw** [comment]
Thanks. 
Adding these two lines

```
                if t==1 and i==no_depths-1:
                    print(U[-1])
```                    

 [here](https://github.com/MindTheGap-ERC/reactive-transport-model-for-limestone-marl-sequences/blob/487ebb7376d804e8ae5a9107b0d33289c57844fb/marlpde/LHeureux_model.py#L350) yields almost the same value:
 
 ```
12.539086320888092
```

as from a breakpoint [here](https://github.com/MindTheGap-ERC/reactive-transport-model-for-limestone-marl-sequences/blob/487ebb7376d804e8ae5a9107b0d33289c57844fb/marlpde/LHeureux_model.py#L112):
```
12.539086322190391
```

I would consider these values the same, when taking into account accuracies in numerical representations.

**Comment by HannoSpreeuw** [comment]
These values come from a value of `Phi = state.data[4][-1]` of `0.9812637484259564`.

What is `phi_bottom = data[3000,4,199]` in your [notebook](https://github.com/MindTheGap-ERC/LMA_utils/blob/main/Track_U_at_bottom.ipynb)?

**Comment by HannoSpreeuw** [comment]
I should add that my run was only over T*, not 30 times T*.

So I have `tmax: int     = Map_Scenario().Tstar` 
cf. [line 161 in parameters.py](https://github.com/MindTheGap-ERC/reactive-transport-model-for-limestone-marl-sequences/blob/487ebb7376d804e8ae5a9107b0d33289c57844fb/marlpde/parameters.py#L161C5-L161C41).

**Comment by EmiliaJarochowska** [comment]
Oh btw if you get annoyed by not being able to reproduce it, you should be able to access the same file by running:

```
import datahugger
dh = datahugger.get("10.17605/OSF.IO/EWS27", "../data")
Solution_path = '../data/21_09_2023_09_29_17/LMAHeureuxPorosityDiff.hdf5' 
Solution = h5py.File(Solution_path, 'r')
data = Solution['data']
```
but you might need to install datahugger. Couldn't come up with a better solution for sharing data files so we can retrieve data from runs.

**Comment by HannoSpreeuw** [comment]
Okay, if the porosities are the same, then one of the constants (presum or rhorat) must differ, there is no other way to explain the difference between 12.54 and 11.97.

**Comment by HannoSpreeuw** [comment]
`datahugger` is a great way to share data. Thanks! I did not know about this tool. I can easily run your notebook after a `!pip install datahugger`.

The difference between 11.97 and 12.54 in U comes from a difference in `rhorat`, which you have calculated to be 1.734, while I have 1.799.

You have 
`rhorat = (rhos_bottom / rhow - 1) * beta / S`
while I have 
`rhorat = (rhos/ rhow - 1) * beta / S`
with an initialization in [parameters.py of `rhos = rhos0`](https://github.com/MindTheGap-ERC/reactive-transport-model-for-limestone-marl-sequences/blob/487ebb7376d804e8ae5a9107b0d33289c57844fb/marlpde/parameters.py#L136), which has been copied from the [original Matlab script](https://github.com/MindTheGap-ERC/LMA-Matlab/blob/f3298a7a9634c5355ec5756a0075b405f471dd49/ScenarioA.m#L24).

So, effectively, in marlpde:
`rhorat = rhorat0`

while your notebook has
`rhos = rhos_bottom`, so `rhorat != rhorat0`.

**Comment by HannoSpreeuw** [comment]
Perhaps the choice `rhos = rhos0` is a consequence of the constant density approximation, described in paragraph 2.5 of L'Heureux:

`We thus set ρ_s equal to a constant defined by the initial composition of the sediment and let ρ_A ≅ ρ_C ≅ ρ_T ≅ ρ_s` . 

Not sure what has been implemented in the FORTRAN code.

**Comment by HannoSpreeuw** [comment]
In the [FORTRAN code, line 137](https://github.com/astro-turing/lheureux/blob/6fd4acc010aad2dbc82d23aac16546ba751beed4/src/lheureux.f#L137) you have:
```
rhosw0=P(20)
```
and for [u(i), line 260](https://github.com/astro-turing/lheureux/blob/6fd4acc010aad2dbc82d23aac16546ba751beed4/src/lheureux.f#L260C14-L260C71):

```
u(i)=k(i)*(1-ph(i))*(rhosw0-1)+P(23)-P(13)*k(0)*(1-ph(0))
```

and [P(20) seems to be set in line 841](https://github.com/astro-turing/lheureux/blob/6fd4acc010aad2dbc82d23aac16546ba751beed4/src/lheureux.f#L841):

```
P(20)=rhos0/rhow
```
so that seems to agree with the Matlab and Python codes, but this interpretation could be too simple. 

@NiklasHohmann How did you get to [`rhos=rhos0;`](https://github.com/MindTheGap-ERC/LMA-Matlab/blob/f3298a7a9634c5355ec5756a0075b405f471dd49/ScenarioA.m#L24)?

**Comment by NiklasHohmann** [comment]
@HannoSpreeuw via the constant density approximation as described by you above

**Comment by EmiliaJarochowska** [comment]
Thanks for this detective work. I believe I made a mistake by disregarding the constant density approximation. I should have noted on why I assumed that `rhos` was to be evaluated for each node. I think maybe because both Fortran and Python codes allow for `rhos` ≠ `rhos0`. E.g. in Fortran we can see that Ivan clearly ran both options, but constant density would be used with the current settings:
```
!  case where all densities are equal to rhos
           P(7)=ccal0
           P(9)=0.
           P(10)=0.
           P(11)=rhos0/rhow
           P(12)=P(11)
           P(27)=1.  
```
https://github.com/astro-turing/lheureux/blob/6fd4acc010aad2dbc82d23aac16546ba751beed4/src/lheureux.f#L853-L859

I'm sorry this costed you a lot of extra work, @HannoSpreeuw. I'll triple-check next time.
Correcting `rhos_bottom` to `rhos0` gives the same values of U as saved in the file so we can merge.
